cpp/src/parquet/column_writer.cc

```diff
@@ -442,10 +442,13 @@ class BufferedPageWriter : public PageWriter {
     // index_page_offset = -1 since they are not supported
     PARQUET_ASSIGN_OR_THROW(int64_t final_position, final_sink_->Tell());
     // dictionary page offset should be 0 iff there are no dictionary pages
-    auto dictionary_page_offset =
-        has_dictionary_pages_ ? pager_->dictionary_page_offset() + final_position : 0;
-    metadata_->Finish(pager_->num_values(), dictionary_page_offset, -1,
-                      pager_->data_page_offset() + final_position,
+    int64_t dictionary_page_offset = 0;
+    int64_t data_page_offset = final_position;
+    if (has_dictionary_pages_) {
+      dictionary_page_offset = pager_->dictionary_page_offset() + final_position;
+      data_page_offset += pager_->data_page_offset();
+    }
+    metadata_->Finish(pager_->num_values(), dictionary_page_offset, -1, data_page_offset,
```

Comment:
I'm not totally following the changes relative to the unit test. 

* The dictionary_page_offset is unchanged 
* When has_dictionary_pages_ is false, data_page_offset is now `final_position` instead of `final_position + pager_->data_page_offset()`. But here `pager_->data_page_offset()` **should be** zero. Why isn't it?

I'm looking at:

```
    PARQUET_ASSIGN_OR_THROW(int64_t start_pos, sink_->Tell());
    if (data_page_offset_ == 0) {
      data_page_offset_ = start_pos;
    }
```

This logic seems wrong. `data_page_offset_` is being set to the offset of the start of the **second** data page. Do you agree? I think that's what we need to fix

Reply:
Yes, @wesm , that's definitely the underlying cause.  That conditional is tripped on second page that looks like it's meant to be part of first page processing. 

I went back and forth on fixing it with the approach you took and the one I took.  The reason I settled on the latter was that it seemed slightly better to scope any changes to the buffered writer since that's where the error manifests.  IOW, I conservatively assumed the the data page offset check in the serialized writer has another purpose besides finding the first page.  But I don't feel strongly about it and if you think changing the serialized writer is safe (or preferable), I'm good with that.  

Just one nit with your patch: there's already a member in the serialized writer called page_ordinal_ that we can check for zero to find the first page rather than introducing a new boolean.  Could be something like:
```
PARQUET_ASSIGN_OR_THROW(int64_t start_pos, sink_->Tell());
if (page_ordinal_ == 0 && data_page_offset_ == 0) {
  data_page_offset_ = start_pos;
}
```
or maybe just:
```
PARQUET_ASSIGN_OR_THROW(int64_t start_pos, sink_->Tell());
if (page_ordinal_ == 0) {
  data_page_offset_ = start_pos;
}
```
Your call.  Let me know and I'll push another commit if necessary.

Reply:
Ah I see. Yes let's use the page ordinal rather than adding the new member. Thank you!